File: 25-reverse-nodes-in-k-group/reverse-nodes-in-k-group.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
ListNode * findKthNode (ListNode * root , int k){
    k--;
    while (k>0 and root!=nullptr){
        k--;
        root=root->next;
    }
    
    return root;
}
void reverseKGroup(ListNode * h){
ListNode * prev=NULL;
ListNode * curr=h;
ListNode * nextPtr=curr;
ListNode * root=nullptr;
    while(curr!=nullptr){
        nextPtr = curr->next;
        curr-> next = prev;
        prev = curr;
        curr= nextPtr;
        
    }
    while(curr!=nullptr){
        cout<<curr->val<<"->";
        curr=curr->next;
    }
    cout<<endl;


}
    ListNode* reverseKGroup(ListNode* head, int k) {



        ListNode * temp = head; 
        ListNode * prev=NULL;
        ListNode * nextt = NULL;
        while(temp !=nullptr){
            ListNode * kthNode = findKthNode(temp,k);
            if(kthNode == NULL){
                // connect simply 
                prev-> next = temp;
                break ;

            }
            nextt = kthNode-> next ;
            kthNode -> next =NULL;
            //reverse k group
            reverseKGroup(temp);

            if (temp == head ){
                head =kthNode;
            }
            else {
                prev-> next = kthNode;
            }
            prev=temp;
            temp=nextt;
        }


        return head;
        
    }
};

File: 25-reverse-nodes-in-k-group/reverse-nodes-in-k-group.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */
function reverse(head: ListNode | null) {
    let prev: ListNode | null = null;
    let curr: ListNode | null = head;
    let nextt: ListNode | null = head;
    while (head != null) {
        nextt = head.next;
        head.next = prev;
        prev = head;
        head = nextt;


    }
    return prev;

}
function reverseKGroup(head: ListNode | null, k: number): ListNode | null {
    let root: ListNode | null = new ListNode(-1);
    let n: number = 0;
    let tail : ListNode | null = null;
    while (head != null) {
        let temp: ListNode | null = head;
        let prev: ListNode | null = head;
        n = 0;
        while (n != k && head != null) {
            prev = head;
            head = head.next;
            n++;
        }
        if(n!=k) {
        tail.next=temp;
        return root.next;}
        prev.next = null;
        let reversedHead = reverse(temp)
        if (root.next == null) root.next = reversedHead;
          if(tail!=null)tail.next = reversedHead;
          tail=temp;
        // console.log("Head ",head!=null?head.val:"()"," Prev ",prev.val," temp->" , temp.val," reversedHead",reversedHead.val);
    }
    return root.next;
};